src/section.rs

```diff
@@ -47,237 +48,365 @@ impl Section {
+    /// Note: there can be multiple section ticks per network tick.
+    pub fn tick(&mut self, params: &Params) -> Vec<Action> {
+        // If we are merging or splitting, resend the message to let
+        // the post-merge/post-split sections receive it.
+        if self.is_merging_or_splitting() {
+            return self.messages
+                .drain(..)
+                .map(|message| Action::Send(message))
+                .collect();
+        }
+
+        let mut actions = Vec::new();
+        let mut relocated_in = false;
 
-        for request in mem::replace(&mut self.requests, Vec::new()) {
+        for message in mem::replace(&mut self.messages, Vec::new()) {
             debug!(
                 "{}: received {}",
                 log::prefix(&self.prefix),
-                log::message(&request)
+                log::message(&message)
             );
 
-            responses.extend(match request {
-                Request::Live(node) => self.handle_live(params, node),
-                Request::Dead(name) => self.handle_dead(params, name),
-                Request::Merge(prefix) => self.handle_merge(params, prefix),
-                Request::RelocateRequest {
-                    src,
-                    dst,
-                    node_name,
-                } => self.handle_relocate_request(params, src, dst, node_name),
-                Request::RelocateAccept { dst, node_name } => {
-                    self.handle_relocate_accept(dst, node_name)
+            match message {
+                Message::RelocateRequest { node_name, target } => {
+                    actions.push(if relocated_in {
+                        Action::Send(Message::RelocateReject { node_name, target })
+                    } else {
+                        self.handle_relocate_request(params, node_name, target)
+                    })
                 }
-                Request::RelocateReject { dst, node_name } => {
-                    self.handle_relocate_reject(dst, node_name)
+                Message::RelocateAccept { node_name, target } => {
+                    actions.extend(self.handle_relocate_accept(node_name, target))
                 }
-                Request::Relocate(node) => self.handle_relocate(params, node),
-            })
+                Message::RelocateReject { node_name, target } => {
+                    actions.extend(self.handle_relocate_reject(params, node_name, target));
+                }
+                Message::RelocateCommit { node, .. } => {
+                    if let Some(action) = self.handle_relocate_commit(params, node) {
+                        relocated_in = true;
+                        actions.push(action);
+                    }
+                }
+                Message::RelocateCancel { node_name, .. } => self.handle_relocate_cancel(node_name),
+                Message::Merge { target, .. } => actions.push(self.handle_merge(params, target)),
+            }
+        }
+
+        if !relocated_in {
```

Comment:
do we also need to check whether the incoming_cache is empty, or the `relocated_in` needs to set to true when receive `RelocateRequest` ?